### PR TITLE
Bootcamp Wk7 Exercise Dmitry

### PR DESCRIPTION
**Pull request by dkisselev-zz**
## Summary

This submission implements an evaluation framework for a fine-tuned LLaMA 3.1-8B model that predicts product prices. The notebook focuses on model inference optimization and hyperparameter tuning of the @ed-donner model and the fine-tuned model created by me.

My fine-tune model was not able to achieve Ed's accuracy, but served as a good example for the optimization technique. That said, the original Ed's model with the technique implemented was improved for $0.62, showing as a viable technique for inference hyperparameter tuning.

## Key Features
1. Model Infrastructure
  - Loads Meta-Llama-3.1-8B base model with QLoRA (4-bit/8-bit quantization)
  - Integrates PEFT adapters from the fine-tuned model
  - Supports loading models from both HuggingFace and Weights & Biases

2. Hyperparameter Optimization
  - Search_K Class: Systematically searches for optimal k value (up to 900) for top-k weighted predictions
  - Analyzes error metrics across different k values
  - Identifies the best k that minimizes the average absolute error
  - Visualizes k vs. error relationship

3. Probability Distribution Analysis
  - Tracks standard deviation of probability distributions across inferences
  - Generates three types of visualizations:
  - K vs. Average Error plot
  - Example probability spreads (min/median/max)
  - Full distribution histogram and line plot

4. Weighted Prediction Strategy
  - Implements normalized weighted averaging using top-k token probabilities
  - Converts token predictions to numeric prices
  - Filters invalid predictions automatically

### Configuration
  - Project: pricer
  - Model Run: Uses Ed's reference model (2024-09-13_13.04.39)
  - Model Run: Uses my model (2025-10-23_23.41.24:v22)
  - Dataset: ed-donner/pricer-data

### Discussion
- The optimization techniques of 4/8 bits during inference can affect the accuracy of the model. Ed's model had the best accuracy as a 4-bit model, and mine was as an 8-bit model. 
- Even though both models were trained on the same dataset, prompt engineering didn't improve Ed's model accuracy but significantly affected mine.
- The weighted average of the prices based on the probabilities helped to increase model performance in both scenarios

### Ed's model performance
- Original Model with $47.29 error
ed4bit-oldprompt
- Optimization shows $46.54 error potential
22-best-k
22_spread

### My model performance
- Original model $53.92 error

- Optimization shows $51.30 error after tuning and optimizing the prompt
21-best-k
21_spread

### CODE REVIEW

**Comment by ed-donner** [comment]
With over 100,000 students taking this course, I can say that your top k analysis is the most interesting analysis that anyone has done from week 7, Dmitry. Hugely enjoyed reviewing this fascinating work.
